config/config_substitute.go

```diff
@@ -5,109 +5,71 @@ import (
-			*value = strings.Replace(*value, getScriptsFolder, scriptFolder, -1)
-			*value = strings.TrimSpace(collections.UnIndent(*value))
-		}
-
-		return value
-	}
-
-	substitute(conf.Uniqueness)
+func (conf *TerragruntConfig) SubstituteAllVariables() {
+	conf.substitute(conf.Uniqueness)
 
 	if roles, ok := conf.AssumeRole.([]string); ok {
 		for i := range roles {
-			substitute(&roles[i])
+			conf.substitute(&roles[i])
 		}
 		conf.AssumeRole = roles
 	}
 
-	if conf.Terraform != nil {
-		for i, extraArgs := range conf.Terraform.ExtraArgs {
-			substitute(&extraArgs.Description)
-			conf.Terraform.ExtraArgs[i] = extraArgs
-		}
-		substitute(&conf.Terraform.Source)
-	}
 	if conf.RemoteState != nil && conf.RemoteState.Config != nil {
 		for key, value := range conf.RemoteState.Config {
 			switch val := value.(type) {
 			case string:
-				conf.RemoteState.Config[key] = *substitute(&val)
+				conf.RemoteState.Config[key] = *conf.substitute(&val)
 			}
 		}
 	}
 
-	substituteHooks := func(hooks HookList) {
-		for i, hook := range hooks {
-			substitute(&hook.Command)
-			substitute(&hook.Description)
-			for i, arg := range hook.Arguments {
-				hook.Arguments[i] = *substitute(&arg)
-			}
-			hooks[i] = hook
-		}
+	if conf.Terraform != nil {
+		conf.substitute(&conf.Terraform.Source)
+	}
+	for i := range conf.ExtraArgs {
+		conf.ExtraArgs[i].substituteVars()
+	}
+	for i := range conf.PreHooks {
+		conf.PreHooks[i].substituteVars()
 	}
-	substituteHooks(conf.PreHooks)
-	substituteHooks(conf.PostHooks)
+	for i := range conf.PostHooks {
+		conf.PostHooks[i].substituteVars()
+	}
+	for i := range conf.ExtraCommands {
+		conf.ExtraCommands[i].substituteVars()
+	}
+	for i := range conf.ImportFiles {
+		conf.ImportFiles[i].substituteVars()
+	}
+	for i := range conf.ImportVariables {
+		conf.ImportVariables[i].substituteVars()
+	}
+}
 
-	for i, command := range conf.ExtraCommands {
-		substitute(&command.Description)
-		substitute(&command.VersionArg)
-		for i, cmd := range command.Commands {
-			command.Commands[i] = *substitute(&cmd)
-		}
-		for i, alias := range command.Aliases {
-			command.Aliases[i] = *substitute(&alias)
-		}
-		for i, arg := range command.Arguments {
-			command.Arguments[i] = *substitute(&arg)
-		}
-		conf.ExtraCommands[i] = command
+// substitute is an helper function to convert string in a configuration structure
+func (conf *TerragruntConfig) substitute(value *string) *string {
+	if value == nil {
+		return nil
 	}
 
-	for i, importer := range conf.ImportFiles {
-		substitute(&importer.Description)
-		substitute(&importer.Source)
-		substitute(&importer.Target)
-		for i, value := range importer.Files {
-			importer.Files[i] = *substitute(&value)
-		}
-		for _, value := range importer.CopyAndRename {
-			substitute(&value.Source)
-			substitute(&value.Target)
-		}
-		conf.ImportFiles[i] = importer
+	*value = SubstituteVars(*value, conf.options)
+	if !conf.options.IgnoreRemainingInterpolation {
```

Comment:
Should be the opposite. If `ignore` is true, it's the last call...

Reply:
I don't think so. This flag is set to false at the end of the evaluation process. And we want to complete these substitution only when all variables have been evaluated.

Reply:
I know that's how it is in the code but it doesn't make sense. If it's the last evaluation, it would make sense for `IgnoreRemainingInterpolation` to be true